_src/_posts/2018-09-07-writing-custom-presto-functions.markdown

```diff
@@ -0,0 +1,108 @@
+You can, however, define the initial parameters of the state via its constructor.
+
+## Implementation
+
+UDFs are implemented inside a [plugin](https://prestodb.io/docs/current/develop/spi-overview.html), separate from the default Presto functions.
+They are exposed to presto by declaring them in [UdfPlugin.java](TODO) and then deploying them as a `.jar` file.
+
+```java
+public class UdfPlugin
+        implements Plugin
+{
+    @Override
+    public Set<Class<?>> getFunctions()
+    {
+        return ImmutableSet.<Class<?>>builder()
+                .add(JSONAggregation.class)
+                .add(YourUDF.class)
+                .build();
+    }
+}
+```
+
+Aggregation functions are trickier to implement than scalar functions, as they have many more moving parts.
+Aside from the `input`, `combine` and `output` functions, you should write a [State](TODO) interface and its auxiliary files.
+If your state uses just basic data types, Presto automatically knows how to construct, serialize and deserialize it.
+Else, you should implement a [Factory](TODO) and a [Serializer](TODO), and link them to the State using Presto's Metadata Annotations.
+
+```java
+@AccumulatorStateMetadata(stateFactoryClass = JSONAggregationStateFactory.class, stateSerializerClass = JSONAggregationStateSerializer.class)
+public interface JSONAggregationState
+        extends AccumulatorState
+{
+    Map<String, Object> getMap();
+
+    void setMap(Map<String, Object> value);
+}
+```
+
+If you'd like to write an UDF that takes different data types as input (_JSON_ and _VARCHAR_, for example), there are two ways of going about it.
+The simple one is to write two (or more) `input` functions that take those data types as parameters.
+This works in our case, but if you need your UDF to be more generic, you can follow PrestoDB's example in their [native functions](https://github.com/prestodb/presto/tree/master/presto-main/src/main/java/com/facebook/presto/operator/aggregation).
+
+## Testing
+
+Once you have your UDF, the sensible thing is to test it. But how do you simulate Presto's inner workings? This is especially important for aggregation functions, given that Presto takes the components of your UDF and uses them inside a black box.
+Thankfully, we have [Docker](https://www.docker.com/)! [Here](TODO) we provide a Makefile and a docker-compose.yml[^2] for getting a Presto server up and running locally, automatically incorporating your UDF.
+
+[^2]: Thanks to [Lewuathe's docker-presto-cluster](https://github.com/Lewuathe/docker-presto-cluster)!
+
+Once it's up, it can be queried like any other Presto cluster, so you can use it to test your UDF.
+
+## Pitfalls
+
+> _"My kingdom for a discernible stack trace!"_
+>> King Richard III
+
+Developing software never is a smooth-sailing kind of deal (where would the fun in that be?), so be prepared to stare into weird looking stack traces.
```

Comment:
Is it into or at?

Reply:
They're slightly different actions, I would argue that "stare into" fits better when you're trying to make sense of a stack trace: https://english.stackexchange.com/a/120603

Reply:
From my point of view you "stare at a screen" so you are "staring at a stack trace". But as you prefer.

Reply:
Into adds a felling of deepness, what I imagine from a weird stack trace that goes on and on into function calls. :man_shrugging: